futurex_openedx_extensions/helpers/roles.py

```diff
@@ -628,4 +636,403 @@ def get_course_access_roles_queryset(  # pylint: disable=too-many-arguments
     else:
         queryset = queryset.filter(org__in=orgs_filter)
 
+    queryset = queryset.annotate(org_lower_case=Lower('org'))
+
     return _apply_special_filters(queryset, course_ids_filter, remove_redundant, exclude_role_type, user_id_distinct)
+
+
+def cache_refresh_course_access_roles() -> None:
+    """Refresh the course access roles cache."""
+    cache.delete(cs.CACHE_NAME_ALL_COURSE_ACCESS_ROLES)
+    get_all_course_access_roles()
+
+
+def delete_course_access_roles(tenant_ids: list[int], user: get_user_model) -> None:
+    """
+    Delete the course access roles for the given tenant IDs and user.
+
+    :param tenant_ids: The tenant IDs to filter on
+    :type tenant_ids: list
+    :param user: The user to filter on
+    :type user: get_user_model
+    """
+    orgs = get_course_org_filter_list(tenant_ids)['course_org_filter_list']
+
+    delete_count, _ = CourseAccessRole.objects.filter(user=user).filter(
+        Q(org__in=orgs) | Q(org=''),
+    ).delete()
+
+    if not delete_count:
+        raise FXCodedException(
+            code=FXExceptionCodes.ROLE_DELETE,
+            message=f'No role found to delete for the user ({user.username}) within the given tenants {tenant_ids}!',
+        )
+
+    cache_refresh_course_access_roles()
+
+
+def _clean_course_access_roles(redundant_hashes: set[str], user: get_user_model, separator: str) -> None:
+    """
+    Clean the redundant course access roles by deleting related records of the given hashes. This function
+    assumes that the input is valid.
+
+    :param redundant_hashes: The redundant hashes to clean
+    :type redundant_hashes: set
+    :param user: The user to filter on
+    :type user: get_user_model
+    :param separator: The separator to split the hashcode
+    """
+    for hashcode in redundant_hashes:
+        try:
+            course_id: str | object = hashcode.split(separator)[2]
+            if course_id == 'None':
+                course_id = CourseKeyField.Empty
+
+            delete_count, _ = CourseAccessRole.objects.filter(
+                user=user,
+                role=hashcode.split(separator)[0],
+                org=hashcode.split(separator)[1],
```

Comment:
Now the hashcode is both a dict key and an information holder i.e. regular class.

```suggestion
                role=hashcode.role,
                org=hashcode.org,
```